pkg/config/interface.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/resource"
+)
+
+// Config defines the interface that project configuration types must follow
+type Config interface {
+	/* Version */
+
+	// GetVersion returns the current project version
+	GetVersion() Version
+
+	/* String fields */
+
+	// GetDomain returns the project domain
+	GetDomain() string
+	// SetDomain sets the project domain
+	SetDomain(domain string) error
+
+	// GetRepository returns the project repository.
+	GetRepository() string
+	// SetRepository sets the project repository
+	SetRepository(repository string) error
+
+	// GetName returns the project name
+	// This method was introduced in project version 3-alpha.
+	GetName() string
+	// SetName sets the project name
+	// This method was introduced in project version 3-alpha.
+	SetName(name string) error
+
+	// GetLayout returns the config layout
+	// This method was introduced in project version 3-alpha.
+	GetLayout() string
+	// SetLayout sets the Config layout
+	// This method was introduced in project version 3-alpha.
+	SetLayout(layout string) error
+
+	/* Boolean fields */
+
+	// IsMultiGroup checks if multi-group is enabled
+	IsMultiGroup() bool
+	// SetMultiGroup enables multi-group
+	SetMultiGroup() error
+	// ClearMultiGroup disables multi-group
+	ClearMultiGroup() error
+
+	// IsComponentConfig checks if component config is enabled
+	// This method was introduced in project version 3-alpha.
+	IsComponentConfig() bool
+	// SetComponentConfig enables component config
+	// This method was introduced in project version 3-alpha.
+	SetComponentConfig() error
+	// ClearComponentConfig disables component config
+	// This method was introduced in project version 3-alpha.
+	ClearComponentConfig() error
+
+	/* Resources */
+
+	// ResourcesLength returns the number of tracked resources
+	ResourcesLength() int
```

Comment:
I do not think that we should have ResourcesLength() in the interface. 
Instead of that, we should've GetResources or Resources which is more generic.
The Length is a go func

Reply:
As slices and underlying arrays are mutable, we can't just return the field. We will need to (deep) copy the whole array. This also includes making a copy of anything that is a pointer. This is a very expensive operation just to compute the length of a slice, that is why `ResourcesLength` is being provided directly. So even if we do includea `GetResources` method, `len(cfg.GetResources())` will be a very bad approach.